android/app/src/main/java/org/jitsi/meet/AudioModeModule.java

```diff
@@ -0,0 +1,440 @@
+ *
+ * Video calls should should use <tt>AudioModeModule.VIDEO_CALL</tt>, which uses the builtin
+ * speaker, earpiece, wired headset or bluetooth headset. The builtin speaker is the default
+ * audio device.
+ *
+ * Before a call has started and after it has ended the <tt>AudioModeModule.DEFAULT</tt> mode
+ * should be used.
+ */
+public class AudioModeModule extends ReactContextBaseJavaModule {
+    /**
+     * Constants representing the audio mode.
+     */
+    private static final int DEFAULT    = 0;
+    private static final int AUDIO_CALL = 1;
+    private static final int VIDEO_CALL = 2;
+
+    /**
+     * React Native module name.
+     */
+    private static final String MODULE_NAME = "AudioMode";
+
+    /**
+     * Tag used when logging messages.
+     */
+    private static final String TAG = MODULE_NAME;
+
+    /**
+     * Audio mode currently in use.
+     */
+    private int currentMode = -1;
+
+    /**
+     * {@link AudioManager} instance used to interact with the Android audio subsystem.
+     */
+    private AudioManager audioManager;
+
+    /**
+     * {@link Handler} for running all operations on the main thread.
+     */
+    private Handler mainThreadHandler;
+
+    /**
+     * {@link Runnable} for running update operation on the main thread.
+     */
+    private Runnable mainThreadRunner;
+
+    /**
+     * {@link BluetoothManager} for detecting Bluetooth device changes in old (< M)
+     * Android versions.
+     */
+    private BluetoothManager bluetoothManager;
+
+    /**
+     * Initializes a new module instance. There shall be a single instance of this module throughout
+     * the lifetime of the application.
+     *
+     * @param reactContext the {@link ReactApplicationContext} where this module is created.
+     */
+    public AudioModeModule(ReactApplicationContext reactContext) {
+        super(reactContext);
+
+        audioManager = ((AudioManager) reactContext.getSystemService(Context.AUDIO_SERVICE));
+        mainThreadHandler = new Handler(Looper.getMainLooper());
+        mainThreadRunner = new Runnable() {
+            @Override
+            public void run() {
+                if (currentMode != -1) {
+                    updateAudioRoute(currentMode);
+                }
+            }
+        };
+        bluetoothManager = new BluetoothManager();
+
+        // Setup runtime device change detection
+        setupAudioRouteChangeDetection();
+    }
+
+    /**
+     * Gets the name for this module, to be used in the React Native bridge.
+     *
+     * @return a string with the module name.
+     */
+    @Override
+    public String getName() {
+        return MODULE_NAME;
+    }
+
+    /**
+     * Gets a mapping with the constants this module is exporting.
+     *
+     * @return a {@link Map} mapping the constants to be exported with their values.
+     */
+    @Override
+    public Map<String, Object> getConstants() {
+        final Map<String, Object> constants = new HashMap<>();
```

Comment:
The advantages of using `final` for local variables, especially in small functions, are questionable.